store-implementation/accumulo-store/src/main/java/uk/gov/gchq/gaffer/accumulostore/MiniAccumuloStore.java

```diff
@@ -0,0 +1,274 @@
+import uk.gov.gchq.gaffer.operation.impl.compare.Max;
+import uk.gov.gchq.gaffer.operation.impl.compare.Min;
+import uk.gov.gchq.gaffer.operation.impl.compare.Sort;
+import uk.gov.gchq.gaffer.operation.impl.export.GetExports;
+import uk.gov.gchq.gaffer.operation.impl.export.set.ExportToSet;
+import uk.gov.gchq.gaffer.operation.impl.export.set.GetSetExport;
+import uk.gov.gchq.gaffer.operation.impl.function.Aggregate;
+import uk.gov.gchq.gaffer.operation.impl.function.Filter;
+import uk.gov.gchq.gaffer.operation.impl.function.Transform;
+import uk.gov.gchq.gaffer.operation.impl.generate.GenerateElements;
+import uk.gov.gchq.gaffer.operation.impl.generate.GenerateObjects;
+import uk.gov.gchq.gaffer.operation.impl.get.GetAdjacentIds;
+import uk.gov.gchq.gaffer.operation.impl.get.GetAllElements;
+import uk.gov.gchq.gaffer.operation.impl.get.GetElements;
+import uk.gov.gchq.gaffer.operation.impl.job.CancelScheduledJob;
+import uk.gov.gchq.gaffer.operation.impl.job.GetAllJobDetails;
+import uk.gov.gchq.gaffer.operation.impl.job.GetJobDetails;
+import uk.gov.gchq.gaffer.operation.impl.job.GetJobResults;
+import uk.gov.gchq.gaffer.operation.impl.join.Join;
+import uk.gov.gchq.gaffer.operation.impl.output.ToArray;
+import uk.gov.gchq.gaffer.operation.impl.output.ToCsv;
+import uk.gov.gchq.gaffer.operation.impl.output.ToEntitySeeds;
+import uk.gov.gchq.gaffer.operation.impl.output.ToList;
+import uk.gov.gchq.gaffer.operation.impl.output.ToMap;
+import uk.gov.gchq.gaffer.operation.impl.output.ToSet;
+import uk.gov.gchq.gaffer.operation.impl.output.ToSingletonList;
+import uk.gov.gchq.gaffer.operation.impl.output.ToStream;
+import uk.gov.gchq.gaffer.operation.impl.output.ToVertices;
+import uk.gov.gchq.gaffer.store.StoreException;
+import uk.gov.gchq.gaffer.store.StoreProperties;
+import uk.gov.gchq.gaffer.store.operation.GetSchema;
+import uk.gov.gchq.gaffer.store.operation.GetTraits;
+import uk.gov.gchq.gaffer.store.operation.add.AddSchemaToLibrary;
+import uk.gov.gchq.gaffer.store.operation.add.AddStorePropertiesToLibrary;
+import uk.gov.gchq.gaffer.store.operation.handler.OperationHandler;
+import uk.gov.gchq.gaffer.store.schema.Schema;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.UUID;
+
+/**
+ * An {@link AccumuloStore} that uses an Accumulo {@link MiniAccumuloCluster} to
+ * provide a {@link Connector}.
+ */
+public class MiniAccumuloStore extends AccumuloStore {
+    private static final String BASE_DIRECTORY = "miniAccumuloStoreTest-";
+    private static final String ROOTPW = "rootPW";
+    private MiniAccumuloCluster miniAccumuloCluster = null;
+    private MiniAccumuloConfig miniAccumuloConfig = null;
+    private Connector miniConnector;
+
+    public StoreProperties setUpTestDB(final StoreProperties properties) throws StoreException {
+        setProperties(properties);
+
+        File targetDir = new File("target");
+        File baseDir;
+        if (targetDir.exists() && targetDir.isDirectory()) {
```

Comment:
If this logic was moved to the pre-initialise method (in combination with the suggestion below), this method would not be needed and could be removed from the Store class also

Reply:
MockAccumulo instances are memory based and are instantiated within the JVM. When they go out of scope they cease to exist. They need no lifecycle handling.
Full Accumulo installations are fully persistent and can be connected to when needed and if the connection is dropped then they can happily continue to exist. They need no lifecycle handling.
MiniAccumulo instances have to be created with details of a directory where the data can reside, and after creation they have to be explicitly started. The filesystem / environment should be (as far as possible) returned to the original state as before the tests the Mini Accumulo instance was created for. This requires that the MiniAccumulo Instance is gracefully closed and the directory structure deleted.
The MiniAccumuloStore thus needs to be explicitly created and started (setUpTestDB).
It also needs to create a set of AccumuloProperties based on the one supplied to create the Accumulo instance, but suitably updated to include the connection details for the zookeeper of the MiniAccumuloStore. It is this set of properties that is returned by setUpTestDB() to be used in all the subsequent tests in the suite.
Additionally we need to create the MiniAccumulo instance in such a way that we can maintain a reference to it so we can call the corresponding tearDownTestDB() to be able to target the correct instance to close it gracefully and then remove the directory structure that the Mini instance used.
The setUpTestDB() is called from the @BeforeAll / @BeforeClass in the test suite in which it is required, and the tearDownTestDB() is called in the @AfterAll / @AfterClass of the same test suite. This ensures that, assuming JUnit is functioning correctly, the MiniAccumulo instance is tidily handled.

If it was moved to the pre-initialise method there would be no handling of the closure and disposal of the instance, and the tests would have no means of accessing the zookeeper details to access the MiniAccumulo instance.

Reply:
Thanks for your responses. To avoid copy and pasting the same thing to each of the comments, as they address similar issues, I'll try and articulate a reply here. 

I feel like the problem lies in the fact that MiniAccumulo is trying to imitate MockAccumulo exactly (by creating the store and then connecting to it). As you've highlighted it would be too inefficient to do this for every single test.

Perhaps it would be better if it was closer aligned to the AccumuloStore. By that, I mean that it merely connects to the store using the Accumulo properties. The setting up and tearing down of the actual Mini Accumulo instance could be done by the tests. This feels like a better solution as you could then create a new store object for each test as you'd only be creating the connection objects, rather than writing to disk.

Reply:
I think you may be assuming that because all the Mini Accumulo Cluster set up and take down code is co-located in the MiniAccumuloStore class together with other Store overridden methods etc. this means that it all the code runs at the same time; it does not.

The issue:
AccumuloStore can simply be instantiated and, provided it has valid zookeeper details in its supplied properties it can work, connecting to the remote Accumulo Server which will already be running and will continue to run after we have finished with it.
MockAccumuloStore can simply be instantiated and will spin up an in-memory Accumulo MockInstance, use it, and can then forget about it. When the references go out of scope the garbage collector will clear it away.
MiniAccumuloStore requires a more complex lifecycle because Mini Accumulo Cluster requires local resources (filestore) to be able to run, and needs explicit starting and stopping.

The solution used:
To run a JUnit test class/suite using MiniAccumuloStore as the Store we need to set up the Mini Accumulo Cluster that the tests will connect to.
This is achieved at the start of the test suite in the @BeforeClass / @BeforeAll annotated method by instantiating a MiniAccumuloStore and calling the MiniAccumuloStore.setUpTestDB() class supplying the StoreProperties already provided in the test. This method:
- creates a Mini Accumulo Cluster with its files under the project's target directory
- starts the cluster
- interrogates the cluster to discover its zookeeper socket details
- clones the originally supplied StoreProperties and updates the zookeeper details in it
- returns the cloned and updated StoreProperties
The returned StoreProperties and the MiniAccumuloStore objects are kept in private attributes in this JUnit instance.
In the subsequent test methods of the JUnit suite the returned StoreProperties are the ones used. All the Graph builds etc will use the normal AccumuloStore mechanisms to access the Mini Accumulo Cluster, instantiating MiniAccumuloStore objects as required, each of which will only actually be a holder of the StoreProperties / AccumuloProperties containing the Mini Accumulo Cluster's zookeeper address details.
When all the test methods have been run then JUnit will run the @AfterClass / @AfterAll annotated method. This will use the MiniAccumuloStore instance kept in the JUnit's private attributes to call its tearDownTestDB() method. This particular MiniAccumuloStore instance contains the MiniAccumuloCluster object and the directory name used for the creation of the Mini Accumulo Cluster so it can use them to stop the cluster and then delete the filestore directory it was using.

In the light of the above outline I hope you can see that the MiniAccumuloStore is operating in a manner far closer to the AccumuloStore than the MockAccumuloStore. In the test methods any Graph build "merely connects to the store using the Accumulo properties" as you would like.

Is it that some of your concerns revolve around the inclusion of the 2 methods (setUpTestDB and tearDownTestDB) in the main Store superclass? I will explain the design thinking behind these.
There are a number of tests that previously used but have no explicit dependency on MockAccumuloStore e.g. in the uk.gov.gchq.gaffer.sparkaccumulo test classes. They only have a set of StoreProperties provided which specified the SingleUseMockAccumuloStore as the store class to be used.
This works OK for AccumuloStore and MockAccumuloStore but as we have seen above MiniAccumuloStore requires the Mini Accumulo Cluster to be set up and taken down again.
The Store setup code added to these test suites' @BeforeClass/@BeforeAll was deliberately written with only uk.gov.gchq.gaffer.store superclass dependencies used.
The test setup reads the StoreProperties, instantiates the declared Store class, and calls the setUpTestDB() method with the supplied StoreParameters against it.
In all implementations other than MiniAccumuloStore this simply uses the Store superclass implementation and returns the properties it was called with. The override of the setUpTestDB() in MiniAccumuloStore sets up the Mini Accumulo Cluster instance under the target directory, starts the instance and returns a new set of properties with the updated zookeeper details.
The returned properties, in either case, are then used in the subsequent test methods in the normal way for Graph building etc.
Once all the test methods have been run the tearDownTestDB() method is called against the store. In the 'default' Store superclass implementation this simply returns. For the MiniAccumuloStore implementation this stops the Mini Accumulo Cluster and deletes the directory the instance was using.
As a result of this approach these tests can now simply have their properties changed to support different stores (AccumuloStore, MockAccumuloStore, MiniAccumuloStore or any of the other store implementations) exactly as they could before. The methods are there for use by other store implementations in the future should the need arise. Federated-store tests may have need of this facility.
In hindsight I do admit that the use of the word 'Test' in the new stub Store method names was ill-advised. Maybe setUpStore and tearDownStore would be more appropriate names?

Reply:
In that case, it's probably better that we use the standard accumulo store for tests that really need to use an accumulo store (Such as the AccumuloStoreITs). For tests that just need an in-memory store, they can just use the MapStore.

Here's an example of how I'd do it: https://github.com/gchq/Gaffer/compare/gh-1836-demo?expand=1

I've only made changes to the AccumuloStoreITs in the demo and the rest of the accumulo store test's need refactoring. While doing it, I found that I didn't really need another Store class as the AccumuloStore works well with the MiniAccumuloCluster.